src/gamemode.c

```diff
@@ -103,70 +106,189 @@ gamemode_class_init (GameModeClass *klass)
 {
 }
 
-/* internal */
-
+/* internal helpers */
 
-/* dbus  */
-static void
-got_gamemode_reply (GObject      *source,
-		    GAsyncResult *res,
-		    gpointer      user_data)
+static gboolean
+gamemode_is_allowed_for_app (const char *app_id, GError **error)
 {
-  g_autoptr(GDBusMethodInvocation) inv = user_data;
+  g_autoptr(GVariant) perms = NULL;
+  g_autoptr(GVariant) data = NULL;
   g_autoptr(GError) err = NULL;
-  g_autoptr(GVariant) val = NULL;
-  int r = -1;
+  const char **stored;
+  gboolean ok;
 
-  val = g_dbus_proxy_call_finish (G_DBUS_PROXY (source), res, &err);
+  ok = xdp_impl_permission_store_call_lookup_sync (get_permission_store (),
+                                                   PERMISSION_TABLE,
+                                                   PERMISSION_ID,
+                                                   &perms,
+                                                   &data,
+                                                   NULL,
+                                                   &err);
 
-  if (val != NULL)
-    g_variant_get (val, "(i)", &r);
-  else
-    g_debug ("call to GameMode failed: %s", err->message);
+  if (!ok)
+    {
+      g_dbus_error_strip_remote_error (err);
+      g_debug ("No gamemode permissions found: %s", err->message);
+      g_clear_error (&err);
+    }
+  else if (perms != NULL && g_variant_lookup (perms, app_id, "^a&s", &stored))
+    {
+      g_autofree char *as_str = NULL;
+      gboolean allowed;
+
+      as_str = g_strjoinv (" ", (char **)stored);
+      g_debug ("Notification gamemode for %s: %s", app_id, as_str);
+
+      allowed = !g_strv_contains (stored, "no");
+
+      if (!allowed)
+	g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+		     "GameMode is not allowed for %s", app_id);
+
+      return allowed;
+    }
+
+  g_debug ("No gamemode permissions stored for %s: allowing", app_id);
+
+  ok = xdp_impl_permission_store_call_set_permission_sync (get_permission_store(),
+							   PERMISSION_TABLE,
+							   TRUE,
+							   PERMISSION_ID,
+							   app_id,
+							   (const char*[]) {"yes", NULL},
+							   NULL,
+							   &err);
```

Comment:
This seems wrong to me. We should not write permissions to the store without asking the user about it.
If we don't want to have UI here (and I think you said we don't), then the permissions should just be read, not written back. UI for it can be added in the control-center application panel.

Reply:
This was me cargo-culting semantics from the [notification](https://github.com/flatpak/xdg-desktop-portal/blob/master/src/notification.c#L171) portal. Happy to remove the write part.

Reply:
Ah, we do that there. hmm. Anyway, I don't think we should here